kindelia/src/main.rs

```diff
@@ -575,19 +575,84 @@ fn statement_from_hex(hex: &str) -> Result<Statement, String> {
+  hosts: Vec<SocketAddr>,
 ) -> Result<(), String> {
-  let f =
-    |client: ApiClient, stmts| async move { client.publish_code(stmts).await };
-  let results = run_on_remote(api_url, stmts, f)?;
-  for (i, result) in results.iter().enumerate() {
-    print!("Transaction #{}: ", i);
-    match result {
-      Ok(()) => println!("PUBLISHED (tx added to mempool)"),
-      Err(err) => {
-        println!("NOT PUBLISHED: {}", err)
-      }
+  // setup tokio runtime and unordered joinset (tasks).
+  let runtime = tokio::runtime::Runtime::new().map_err(|e| e.to_string())?;
+  let mut tasks = tokio::task::JoinSet::new();
+
+  let client = ApiClient::new(api_url, None).map_err(|e| e.to_string())?;
+
+  let peer_urls: Vec<String> = if hosts.is_empty() {
+    // obtain list of active peers known to "our" node.
+    let prom = async move { client.get_peers::<NC>(false).await };
+    let peers = runtime.block_on(prom)?;
+    let mut urls: Vec<String> = peers
+      .iter()
+      .map(|p| match p.address {
+        Address::IPv4 { val0, val1, val2, val3, port: _ } => {
+          // strips the port, so port 80 is assumed/default.
+          // note:  we just assume/guess that this host has
+          //        a webservice running on port 80. (gross)
+          //        this is a fragile and temporary hack until
+          //        code is relayed properly by p2p nodes.
+          format!("http://{}.{}.{}.{}", val0, val1, val2, val3)
+        }
+      })
+      .collect();
+    // add api_url if not present in peers list.
+    if !urls.iter().any(|x| x == api_url) {
+      urls.push(api_url.to_string());
     }
+    urls
+  } else {
+    hosts.iter().map(|h| format!("http://{}", h)).collect()
+  };
+
+  let stmts_hex: Vec<HexStatement> =
+    stmts.into_iter().map(|s| s.into()).collect();
+
+  for peer_url in peer_urls.into_iter() {
+    // these are move'd into the spawned task, so they must be
+    // created for each iteration.
+    let client = ApiClient::new(peer_url, None).map_err(|e| e.to_string())?;
+    let stmts_hex = stmts_hex.clone();
+
+    // spawn a new task for contacting each peer.  Because it is
+    // spawn'd, the task should begin executing immediately.
+    tasks.spawn_on(
+      async move {
+        let results = match client.publish_code(stmts_hex.clone()).await {
+          Ok(r) => r,
+          Err(e) => {
+            println!("NOT PUBLISHED to {}. ({})", *client, e);
+            return Err(e);
+          }
+        };
+        for (i, result) in results.iter().enumerate() {
+          print!("Transaction #{}: ", i);
+          match result {
+            Ok(_) => {
+              println!("PUBLISHED to {} (tx added to mempool)", *client)
+            }
+            Err(err) => {
+              println!("NOT PUBLISHED to {}: {}", *client, err)
+            }
+          }
+        }
+        Ok(())
+      },
+      runtime.handle(),
+    );
   }
+  // wait for all tasks to complete.
+  runtime.block_on(join_all(tasks))
```

Comment:
I think this whole part can be done using `futures::join_all` function:
```rust
let tasks = peers_url.into_iter().map(|url| {
  let client = new_client(url)
  ...
})

block_on(join_all(tasks))
```
But for do this the futures crate should be added for this crate, and I don't know if its worth

Reply:
This way would create the client inside the task and, in my opinion, is more readable (although 90% of the repo's code aren't anyway; i hope to change this)

Reply:
yeah, I did it this way to avoid introducing new dep(s).  My feeling is futures::join_all is not worth it, but I can make the change if you want.

Reply:
No, it is not necessary